In [1]:
from ml_enhance import QuantumFPFileLoader, QFPFeatureEngineer, ConformerAggregator, QuantumFPDatasetBuilder, RDKitFeatureCalculator
from pathlib import Path
from rdkit import Chem
import pandas as pd
import numpy as np
import json
import os

In [2]:
OUTPUT_PATH = Path("../data/QuantumFP/QFP_output")

loader = QuantumFPFileLoader(OUTPUT_PATH)
output_files = loader.list_output_files()
output_files_numbers: list[Path] = [int(path.name.split('.')[0].split('_')[-1]) for path in output_files]

In [3]:
print(output_files)

[WindowsPath('../data/QuantumFP/QFP_output/internship_smiles2_0.json.gz'), WindowsPath('../data/QuantumFP/QFP_output/internship_smiles2_1.json.gz'), WindowsPath('../data/QuantumFP/QFP_output/internship_smiles2_10.json.gz'), WindowsPath('../data/QuantumFP/QFP_output/internship_smiles2_100.json.gz'), WindowsPath('../data/QuantumFP/QFP_output/internship_smiles2_1000.json.gz'), WindowsPath('../data/QuantumFP/QFP_output/internship_smiles2_1001.json.gz'), WindowsPath('../data/QuantumFP/QFP_output/internship_smiles2_1002.json.gz'), WindowsPath('../data/QuantumFP/QFP_output/internship_smiles2_1003.json.gz'), WindowsPath('../data/QuantumFP/QFP_output/internship_smiles2_1004.json.gz'), WindowsPath('../data/QuantumFP/QFP_output/internship_smiles2_1005.json.gz'), WindowsPath('../data/QuantumFP/QFP_output/internship_smiles2_1006.json.gz'), WindowsPath('../data/QuantumFP/QFP_output/internship_smiles2_1007.json.gz'), WindowsPath('../data/QuantumFP/QFP_output/internship_smiles2_1008.json.gz'), Windows

In [4]:
# builder = QuantumFPDatasetBuilder(OUTPUT_PATH)
# df, errors = builder.build_dataset(multiprocess=True)

In [5]:
df = pd.read_json('../data/temp_df.json', orient='split', compression='infer')

In [6]:
df.info()

<class 'pandas.DataFrame'>
Index: 8837 entries, 0 to 8836
Data columns (total 56 columns):
 #   Column                                       Non-Null Count  Dtype  
---  ------                                       --------------  -----  
 0   smiles                                       8837 non-null   str    
 1   energy                                       8802 non-null   float64
 2   atomization_energy                           8802 non-null   float64
 3   homo_lumo_gap                                8802 non-null   float64
 4   ionization_energy                            8802 non-null   float64
 5   electron_affinity                            8802 non-null   float64
 6   chemical_potential                           8802 non-null   float64
 7   molecular_dipole_norm                        8802 non-null   float64
 8   molecular_quadrupole_principal_invariant_2   8802 non-null   float64
 9   molecular_quadrupole_principal_invariant_3   8802 non-null   float64
 10  molecular_polari

In [7]:
df["sterimol_Bmax"].to_list()

[5.2549684348,
 3.9595541428,
 4.4360733656,
 4.793257362,
 6.1256587148,
 6.29774399,
 4.8100473134,
 6.3564550272,
 5.2518393449,
 11.86262292,
 5.0792114122,
 4.7045870618,
 6.8048438193,
 5.8587499313,
 10.8269028928,
 4.8864851556,
 5.9725200575,
 4.5967830747,
 5.7324031741,
 8.691689995,
 9.8201800529,
 4.3853416268,
 4.9469555557,
 4.5719543809,
 5.3866938148,
 4.0627224569,
 6.2352985946,
 8.6658057149,
 6.4716246706,
 4.9151133483,
 9.0941090057,
 4.3870715462,
 4.6603705633,
 4.4100004645,
 4.6441328752,
 5.9434760281,
 5.1394050901,
 7.465449549,
 4.2990605962,
 4.5653744534000005,
 7.2272078774,
 5.0787288267,
 4.4379160491,
 5.0527038669,
 7.1800193071,
 5.4038423467,
 5.2376922683,
 4.4741636521,
 4.5978247075,
 5.5658949131,
 5.5113058008,
 11.1814398819,
 8.9934952264,
 7.8344778665,
 6.1062447341,
 3.580910169,
 7.3690720885,
 9.0868252252,
 4.6195781225,
 5.7273386523,
 5.3923588819,
 10.3729125268,
 4.7945604287,
 4.8589533566,
 5.6001188213,
 6.2482014849,
 4.99829

In [8]:
# df.to_json('../data/temp_df.json', orient='split', compression='infer', index=True)

In [9]:
df[df["smiles"]=="[F:1][C:2]([F:3])([F:4])[F:5]"]

,smiles,energy,atomization_energy,homo_lumo_gap,ionization_energy,electron_affinity,chemical_potential,molecular_dipole_norm,molecular_quadrupole_principal_invariant_2,molecular_quadrupole_principal_invariant_3,...,avg_partial_charge,avg_atomic_dipole_norm,avg_atomic_charge_dipole_interaction,avg_bond_stiffness,avg_overlap_integral,avg_bond_length,avg_atomic_dipole_dipole_interaction,avg_nuclear_repulsion,avg_atomic_charge_quadrupole_interaction,avg_bond_energy
105,[F:1][C:2]([F:3])([F:4])[F:5],-21.078563,0.803077,0.66204,0.74286,-0.235545,-0.259115,0.000124,-1.810000e-08,0.0,...,0.0,0.240626,0.01759,0.671279,0.141857,2.483285,0.009155,10.944868,0.005217,NaN


In [10]:
df.isnull().sum()

smiles                                           0
energy                                          35
atomization_energy                              35
homo_lumo_gap                                   35
ionization_energy                               35
electron_affinity                               35
chemical_potential                              35
molecular_dipole_norm                           35
molecular_quadrupole_principal_invariant_2      35
molecular_quadrupole_principal_invariant_3      35
molecular_polarizability_mean                   35
molecular_polarizability_anisotropy             35
enthalpy                                        35
zero_point_energy                               35
radius_of_gyration                              35
molecular_volume                                35
sterimol_L                                      35
sterimol_Bmin                                   39
sterimol_Bmax                                   39
molecular_sasa                 

In [11]:
null_df = df[df.isnull().any(axis=1)].reset_index()

In [12]:
null_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 177 entries, 0 to 176
Data columns (total 57 columns):
 #   Column                                       Non-Null Count  Dtype  
---  ------                                       --------------  -----  
 0   index                                        177 non-null    int64  
 1   smiles                                       177 non-null    str    
 2   energy                                       142 non-null    float64
 3   atomization_energy                           142 non-null    float64
 4   homo_lumo_gap                                142 non-null    float64
 5   ionization_energy                            142 non-null    float64
 6   electron_affinity                            142 non-null    float64
 7   chemical_potential                           142 non-null    float64
 8   molecular_dipole_norm                        142 non-null    float64
 9   molecular_quadrupole_principal_invariant_2   142 non-null    float64
 10  molecular_qua

In [13]:
null_df[null_df["sterimol_Bmax"].isna() & null_df["energy"].notna()]

,index,smiles,energy,atomization_energy,homo_lumo_gap,ionization_energy,electron_affinity,chemical_potential,molecular_dipole_norm,molecular_quadrupole_principal_invariant_2,...,avg_partial_charge,avg_atomic_dipole_norm,avg_atomic_charge_dipole_interaction,avg_bond_stiffness,avg_overlap_integral,avg_bond_length,avg_atomic_dipole_dipole_interaction,avg_nuclear_repulsion,avg_atomic_charge_quadrupole_interaction,avg_bond_energy
44,1266,[S:1]=[Fe:2],-6.290732,0.195666,0.009650,0.423603,0.178285,-0.298930,2.873166,-2.955365,...,1.000000e-10,0.558620,0.065998,0.503757,0.559439,3.735073,0.007325,58.937993,-0.007910,NaN
47,1308,[O:1]=[Mn:2],-6.779702,0.390915,0.047629,0.457364,0.153677,-0.296536,3.026655,-8.978168,...,0.000000e+00,1.303515,0.140346,0.861228,0.450179,2.790336,0.250665,37.929284,-0.042319,NaN
67,1645,[O:1]=[Co:2],-7.563763,0.286841,0.009911,0.492495,0.199619,-0.335602,2.494365,-0.764242,...,0.000000e+00,0.683334,0.121129,0.879061,0.461955,2.837714,0.018588,40.279705,-0.089018,NaN
175,8814,[O:1]=[Y:2],-5.217238,0.252253,0.047072,0.539219,0.156817,-0.349591,2.199255,-32.989183,...,-0.000000e+00,0.206345,0.029041,0.186508,0.346850,3.565589,-0.011236,46.304641,0.107557,NaN


In [22]:
output_files_numbers[1645]

2483

In [15]:
file_numbers = [output_files_numbers[idx] for idx in null_df["index"].to_list()]
file_numbers

[106,
 1083,
 1092,
 1106,
 1116,
 114,
 1167,
 1184,
 1198,
 1218,
 1245,
 1259,
 1272,
 1274,
 1276,
 1278,
 1280,
 1282,
 1286,
 1291,
 1292,
 1300,
 1305,
 1325,
 1333,
 1343,
 1344,
 1355,
 1396,
 1406,
 141,
 1515,
 1523,
 154,
 1572,
 1707,
 1807,
 188,
 1915,
 1925,
 1979,
 1993,
 2091,
 2131,
 2140,
 2146,
 2171,
 2179,
 2183,
 2185,
 2220,
 2299,
 2314,
 2369,
 2370,
 2371,
 2373,
 2377,
 2388,
 2390,
 2397,
 2434,
 2442,
 2443,
 2446,
 2464,
 247,
 2483,
 2501,
 2504,
 2510,
 2515,
 2531,
 2543,
 2550,
 2567,
 2577,
 2578,
 2591,
 2592,
 2593,
 2698,
 2702,
 2703,
 2704,
 2785,
 2786,
 290,
 2992,
 3039,
 3101,
 3168,
 3259,
 3263,
 3374,
 3377,
 346,
 3471,
 3570,
 3575,
 3612,
 3700,
 3825,
 3841,
 3897,
 3899,
 3903,
 3930,
 4324,
 4334,
 4347,
 4349,
 4350,
 449,
 456,
 4574,
 4705,
 4799,
 48,
 485,
 489,
 5084,
 5085,
 5127,
 5174,
 5200,
 5201,
 5473,
 5498,
 5574,
 5576,
 5634,
 5635,
 5787,
 5896,
 5898,
 5913,
 5923,
 5937,
 6072,
 6156,
 620,
 6297,
 636,
 645,
 6

In [16]:
# null_df[null_df["index"]==158].info()

In [23]:
# for df in loader.stream_conformer_dataframe(OUTPUT_PATH / Path("internship_smiles2_114.json.gz")):
for df in loader.stream_conformer_dataframe(OUTPUT_PATH / Path("internship_smiles2_2483.json.gz")):
    tdf = df

# print(tdf.info())
tdf["sterimol_Bmax"]

0    2
Name: sterimol_Bmax, dtype: Int64

In [24]:
tdf[tdf.isnull().any(axis=1)].reset_index()

,index,energy,atomization_energy,homo_lumo_gap,ionization_energy,electron_affinity,chemical_potential,molecular_dipole,molecular_dipole_norm,molecular_quadrupole,...,bond_energy,bond_length,bond_stiffness,overlap_integral,nuclear_repulsion,atomic_charge_dipole_interaction,atomic_charge_quadrupole_interaction,atomic_dipole_dipole_interaction,original_smiles,output_smiles


In [25]:
engineer = QFPFeatureEngineer(300.0)
cleaned_tdf = engineer.clean_features(tdf)
aggregator = ConformerAggregator()
processed_tdf = aggregator.thermal_average(cleaned_tdf)

processed_tdf

C:\Users\thibo\OneDrive\Bureaublad\Enhancing-aqueous-solubility-prediction-models\ml_enhance\qfp_processing\feature_engineering.py:156: RuntimeWarning: Mean of empty slice
  return np.array([x[2] for x in interaction_feature]).mean()
c:\Users\thibo\OneDrive\Bureaublad\Enhancing-aqueous-solubility-prediction-models\.venv\Lib\site-packages\numpy\_core\_methods.py:142: RuntimeWarning: invalid value encountered in scalar divide
  ret = ret.dtype.type(ret / rcount)


smiles                                         [O:1]=[Co:2]
energy                                            -7.563763
atomization_energy                                 0.286841
homo_lumo_gap                                      0.009911
ionization_energy                                  0.492495
electron_affinity                                  0.199619
chemical_potential                                -0.335602
molecular_dipole_norm                              2.494365
molecular_quadrupole_principal_invariant_2        -0.764242
molecular_quadrupole_principal_invariant_3        -0.257155
molecular_polarizability_mean                    -41.003243
molecular_polarizability_anisotropy              592.809366
enthalpy                                          -6.190301
zero_point_energy                                  1.373461
radius_of_gyration                                 0.504338
molecular_volume                                 266.626638
sterimol_L                              

In file 114, has empty bond_energy which leads to something wrong in the engineer

In [20]:
for t in loader.stream_conformer_dataframe(OUTPUT_PATH / "internship_smiles2_0.json.gz"):
    print(t["gibbs_free_energy"])

0    [[200.0, 42.6861593912284], [300.0, 30.8223823...
Name: gibbs_free_energy, dtype: object
